cclib/method/ddec.py

```diff
@@ -0,0 +1,243 @@
+            self.logger.info("Calculating charge densities on the provided empty grid.")
+            if len(self.data.mocoeffs) == 1:
+                self.chgdensity = electrondensity_spin(
+                    self.data, self.volume, [self.data.mocoeffs[0][: self.data.homos[0]]]
+                )
+                self.chgdensity.data *= 2
+            else:
+                self.chgdensity = electrondensity_spin(
+                    self.data,
+                    self.volume,
+                    [
+                        self.data.mocoeffs[0][: self.data.homos[0]],
+                        self.data.mocoeffs[1][: self.data.homos[1]],
+                    ],
+                )
+        # If charge densities are provided beforehand, log this information
+        # `Volume` object does not contain (nor rely on) information about the constituent atoms.
+        else:
+            self.logger.info("Using charge densities from the provided Volume object.")
+            self.chgdensity = self.volume
+
+        # STEP 1
+        # Carry out step 1 of DDEC6 algorithm [Determining ion charge value]
+        # Refer to equations 49-57 in doi: 10.1039/c6ra04656h
+        self.calculate_refcharges()
+
+    def calculate_refcharges(self):
+        # Generator object to iterate over the grid
+        xshape, yshape, zshape = self.chgdensity.data.shape
+        atoms = len(self.data.atomnos)
+        indices = (
+            (i, x, y, z)
+            for i in range(atoms)
+            for x in range(xshape)
+            for y in range(yshape)
+            for z in range(zshape)
+        )
+
+        stockholder_w = numpy.zeros((atoms, xshape, yshape, zshape))
+        localized_w = numpy.zeros((atoms, xshape, yshape, zshape))
+
+        for atomi, xindex, yindex, zindex in indices:
+            # Distance of the grid from atom grid
+            dist_r = self._cartesian_dist(
+                self.data.atomcoords[-1][atomi],
+                self.chgdensity.coordinates([xindex, yindex, zindex]),
+            )
+            closest_r_index = numpy.abs(self.radial_grid_r[atomi] - dist_r).argmin()
+
+            stockholder_w[atomi][xindex][yindex][zindex] = self.proatom_density[atomi][
+                closest_r_index
+            ]
+
+        localized_w = numpy.power(stockholder_w, 4)
+
+        stockholder_bigW = numpy.sum(stockholder_w, axis=0)
+        localized_bigW = numpy.sum(localized_w, axis=0)
+
+        self.logger.info("Creating first reference charges.")
+        self.refcharges = numpy.zeros((atoms))
+        self._localizedcharges = numpy.zeros((atoms))
+        self._stockholdercharges = numpy.zeros((atoms))
+
+        for atomi in range(atoms):
+            chargedensity = copy.deepcopy(self.chgdensity)
+            localized_weight = localized_w[atomi] / localized_bigW
+            chargedensity.data *= localized_weight
+            self._localizedcharges[atomi] = self.data.atomnos[atomi] - chargedensity.integrate()
```

Comment:
` self._localizedcharges[atomi] = self.data.atomnos[atomi] - (self.chgdensity.integrate()*localized_weight)` should do what you want and remove the deepcopy. In a `Volume`object, the sum of the grid data is multiplied by the spacings in x,y,z. Pulling the weighting out of the sum is fine so we can do the above. This prevents the need for a new function.

You may have already checked this out, but in case you didn't!

Reply:
I could be mistaken, but since `localized_weight` is three-dimensional (on x, y, z) and `self.chgdensity.integrate()` returns a float, the dimensions would not match? Let me know if I misunderstood your comment. Thanks!

Reply:
Oh you are right. I wasn't thinking about the shape of localized_weight. I wanted to see if there was a way around adding the function to volume, but it is necessary.

Reply:
#917 should take care of this issue. I'll modify this PR soon based on 917.